File: lib/spells/effects/Obstacle.cpp
/*
 * Obstacle.cpp, part of VCMI engine
 *
 * Authors: listed in file AUTHORS in main folder
 *
 * License: GNU General Public License v2.0 or later
 * Full text of license available in license.txt file, in main folder
 *
 */
#include "StdInc.h"

#include "Obstacle.h"

#include "Registry.h"
#include "../ISpellMechanics.h"

#include "../../battle/IBattleState.h"
#include "../../battle/CBattleInfoCallback.h"
#include "../../networkPacks/PacksForClientBattle.h"
#include "../../serializer/JsonSerializeFormat.h"
#include "../../CRandomGenerator.h"

VCMI_LIB_NAMESPACE_BEGIN

namespace spells
{
namespace effects
{

using RelativeShape = std::vector<std::vector<BattleHex::EDir>>;

static void serializeRelativeShape(JsonSerializeFormat & handler, const std::string & fieldName, RelativeShape & value)
{
	static const std::vector<std::string> EDirMap =
	{
		"TL",
		"TR",
		"R",
		"BR",
		"BL",
		"L",
	};

	{
		JsonArraySerializer outer = handler.enterArray(fieldName);
		outer.syncSize(value, JsonNode::JsonType::DATA_VECTOR);

		for(size_t outerIndex = 0; outerIndex < outer.size(); outerIndex++)
		{
			JsonArraySerializer inner = outer.enterArray(outerIndex);
			inner.syncSize(value.at(outerIndex), JsonNode::JsonType::DATA_STRING);

			for(size_t innerIndex = 0; innerIndex < inner.size(); innerIndex++)
			{
				std::string temp;

				if(handler.saving)
				{
					auto index = value.at(outerIndex).at(innerIndex);
					if (index < EDirMap.size())
						temp = EDirMap[index];
				}

				inner.serializeString(innerIndex, temp);

				if(!handler.saving)
				{
					value.at(outerIndex).at(innerIndex) = static_cast<BattleHex::EDir>(vstd::find_pos(EDirMap, temp));
				}
			}
		}
	}

	if(!handler.saving)
	{
		if(value.empty())
			value.emplace_back();

		if(value.back().empty())
			value.back().emplace_back(BattleHex::EDir::NONE);
	}
}

void ObstacleSideOptions::serializeJson(JsonSerializeFormat & handler)
{
	serializeRelativeShape(handler, "shape", shape);
	serializeRelativeShape(handler, "range", range);

	handler.serializeStruct("appearSound", appearSound);
	handler.serializeStruct("appearAnimation", appearAnimation);
	handler.serializeStruct("animation", animation);

	handler.serializeInt("offsetY", offsetY);
}

void Obstacle::adjustAffectedHexes(std::set<BattleHex> & hexes, const Mechanics * m, const Target & spellTarget) const
{
	EffectTarget effectTarget = transformTarget(m, spellTarget, spellTarget);

	const ObstacleSideOptions & options = sideOptions.at(m->casterSide);

	for(auto & destination : effectTarget)
	{
		for(const auto & trasformation : options.shape)
		{
			BattleHex hex = destination.hexValue;

			for(auto direction : trasformation)
				hex.moveInDirection(direction, false);

			if(hex.isValid())
				hexes.insert(hex);
		}
	}
}

bool Obstacle::applicable(Problem & problem, const Mechanics * m) const
{
	if(hidden && !hideNative && m->battle()->battleHasNativeStack(m->battle()->otherSide(m->casterSide)))
		return m->adaptProblem(ESpellCastProblem::NO_APPROPRIATE_TARGET, problem);

	return LocationEffect::applicable(problem, m);
}

bool Obstacle::applicable(Problem & problem, const Mechanics * m, const EffectTarget & target) const
{
	if(!m->isMassive())
	{
		const bool requiresClearTiles = m->requiresClearTiles();
		const ObstacleSideOptions & options = sideOptions.at(m->casterSide);

		if(target.empty())
			return noRoomToPlace(problem, m);

		for(const auto & destination : target)
		{
			for(const auto & trasformation : options.shape)
			{
				BattleHex hex = destination.hexValue;
				for(auto direction : trasformation)
					hex.moveInDirection(direction, false);

				if(!isHexAvailable(m->battle(), hex, requiresClearTiles))
					return noRoomToPlace(problem, m);
			}
		}
	}

	return LocationEffect::applicable(problem, m, target);
}

EffectTarget Obstacle::transformTarget(const Mechanics * m, const Target & aimPoint, const Target & spellTarget) const
{
	const ObstacleSideOptions & options = sideOptions.at(m->casterSide);

	EffectTarget ret;

	if(!m->isMassive())
	{
		for(const auto & spellDestination : spellTarget)
		{
			for(const auto & rangeShape : options.range)
			{
				BattleHex hex = spellDestination.hexValue;

				for(auto direction : rangeShape)
					hex.moveInDirection(direction, false);

				ret.emplace_back(hex);
			}
		}
	}

	return ret;
}

void Obstacle::apply(ServerCallback * server, const Mechanics * m, const EffectTarget & target) const
{
	if(patchCount > 0)
	{
		std::vector<BattleHex> availableTiles;
		auto insertAvailable = [&m](const BattleHex & hex, std::vector<BattleHex> & availableTiles)
		{
			if(isHexAvailable(m->battle(), hex, true))
				availableTiles.push_back(hex);
		};

		if(m->isMassive())
			for(int i = 0; i < GameConstants::BFIELD_SIZE; i++)
				insertAvailable(BattleHex(i), availableTiles);
		else
			for(const auto & destination : target)
				insertAvailable(destination.hexValue, availableTiles);

		RandomGeneratorUtil::randomShuffle(availableTiles, *server->getRNG());
		const int patchesToPut = std::min(patchCount, static_cast<int>(availableTiles.size()));
		EffectTarget randomTarget;
		randomTarget.reserve(patchesToPut);
		for(int i = 0; i < patchesToPut; i++)
			randomTarget.emplace_back(availableTiles.at(i));
		placeObstacles(server, m, randomTarget);
		return;
	}

	placeObstacles(server, m, target);
}

void Obstacle::serializeJsonEffect(JsonSerializeFormat & handler)
{
	handler.serializeBool("hidden", hidden);
	handler.serializeBool("passable", passable);
	handler.serializeBool("trap", trap);
	handler.serializeBool("removeOnTrigger", removeOnTrigger);
	handler.serializeBool("hideNative", hideNative);

	handler.serializeInt("patchCount", patchCount);
	handler.serializeInt("turnsRemaining", turnsRemaining, -1);
	handler.serializeId("triggerAbility", triggerAbility, SpellID::NONE);

	handler.serializeStruct("attacker", sideOptions.at(BattleSide::ATTACKER));
	handler.serializeStruct("defender", sideOptions.at(BattleSide::DEFENDER));
}

bool Obstacle::isHexAvailable(const CBattleInfoCallback * cb, const BattleHex & hex, const bool mustBeClear)
{
	if(!hex.isAvailable())
		return false;

	if(!mustBeClear)
		return true;

	if(cb->battleGetUnitByPos(hex, true))
		return false;

	auto obst = cb->battleGetAllObstaclesOnPos(hex, false);

	for(auto & i : obst)
		if(i->obstacleType != CObstacleInstance::MOAT)
			return false;

	if(cb->battleGetSiegeLevel() != 0)
	{
		EWallPart part = cb->battleHexToWallPart(hex);

		if(part == EWallPart::INVALID)
			return true;//no fortification here
		else if(part == EWallPart::INDESTRUCTIBLE_PART_OF_GATE)
			return true; // location accessible
		else if(part == EWallPart::INDESTRUCTIBLE_PART)
			return false;//indestructible part (cant be checked by battleGetWallState)
		else if(part == EWallPart::BOTTOM_TOWER || part == EWallPart::UPPER_TOWER)
			return false;//destructible, but should not be available
		else if(cb->battleGetWallState(part) != EWallState::DESTROYED && cb->battleGetWallState(part) != EWallState::NONE)
			return false;
	}

	return true;
}

bool Obstacle::noRoomToPlace(Problem & problem, const Mechanics * m)
{
	MetaString text;
	text.appendLocalString(EMetaText::GENERAL_TXT, 181);//No room to place %s here
	text.replaceRawString(m->getSpellName());
	problem.add(std::move(text));
	return false;
}

void Obstacle::placeObstacles(ServerCallback * server, const Mechanics * m, const EffectTarget & target) const
{
	const ObstacleSideOptions & options = sideOptions.at(m->casterSide);

	BattleObstaclesChanged pack;
	pack.battleID = m->battle()->getBattle()->getBattleID();

	auto all = m->battle()->battleGetAllObstacles(BattlePerspective::ALL_KNOWING);

	int obstacleIdToGive = 1;
	for(auto & one : all)
		if(one->uniqueID >= obstacleIdToGive)
			obstacleIdToGive = one->uniqueID + 1;

	for(const Destination & destination : target)
	{
		SpellCreatedObstacle obstacle;
		obstacle.uniqueID = obstacleIdToGive++;
		obstacle.pos = destination.hexValue;
		obstacle.obstacleType = CObstacleInstance::SPELL_CREATED;
		obstacle.ID = m->getSpellIndex();

		obstacle.turnsRemaining = turnsRemaining;
		obstacle.casterSpellPower = m->getEffectPower();
		obstacle.spellLevel = m->getEffectLevel();//todo: level of indirect effect should be also configurable
		obstacle.casterSide = m->casterSide;

		obstacle.nativeVisible = !hideNative;
		obstacle.hidden = hidden;
		obstacle.passable = passable;
		obstacle.trigger = triggerAbility;
		obstacle.trap = trap;
		obstacle.removeOnTrigger = removeOnTrigger;

		obstacle.appearSound = options.appearSound;
		obstacle.appearAnimation = options.appearAnimation;
		obstacle.animation = options.animation;

		obstacle.animationYOffset = options.offsetY;

		obstacle.customSize.clear();
		obstacle.customSize.reserve(options.shape.size());

		for(const auto & shape : options.shape)
		{
			BattleHex hex = destination.hexValue;

			for(auto direction : shape)
				hex.moveInDirection(direction, false);

			obstacle.customSize.emplace_back(hex);
		}

		pack.changes.emplace_back();
		obstacle.toInfo(pack.changes.back());
	}

	if(!pack.changes.empty())
		server->apply(&pack);
}

}
}

VCMI_LIB_NAMESPACE_END

File: scripts/obstacle.lua
local Base = require("spellEffect")
local Script = setmetatable({}, {__index = Base})
Script.__index = Script

local DIRECTION_METHOD =
{
	TL = "copyToNorthWest",
	TR = "copyToNorthEast",
	R  = "copyToEast",
	BR = "copyToSouthEast",
	BL = "copyToSouthWest",
	L  = "copyToWest",
}

local function applyShape(hex, shape)
	local result = hex
	if shape == nil then return result end
	for _, dir in ipairs(shape) do
		local method = DIRECTION_METHOD[dir]
		if method then
			result = result[method](result)
		end
	end
	return result
end

local function sideOptions(self, side)
	if side == ENUM.BattleSide.attacker then
		return self.attacker or {}
	end
	return self.defender or {}
end

local function otherSide(side)
	if side == ENUM.BattleSide.attacker then return ENUM.BattleSide.defender end
	return ENUM.BattleSide.attacker
end

local function shapesFor(opts)
	local shapes = opts.shape
	if shapes == nil or #shapes == 0 then return { {} } end
	return shapes
end

local function rangesFor(opts)
	local ranges = opts.range
	if ranges == nil or #ranges == 0 then return { {} } end
	return ranges
end

local function isHexAvailable(battle, hex, mustBeClear)
	if not hex:isAvailable() then return false end
	if not mustBeClear then return true end

	if battle:getUnitByPos(hex, true) ~= nil then return false end

	for _, o in ipairs(battle:getObstaclesOnPos(hex, false)) do
		if o:getObstacleType() ~= ENUM.ObstacleType.moat then
			return false
		end
	end

	if battle:hasFortifications() then
		local part = battle:hexToWallPart(hex)
		if part == ENUM.WallPart.invalid then return true end
		if part == ENUM.WallPart.indestructiblePart or part == ENUM.WallPart.indestructibleGate then return false end
		if part == ENUM.WallPart.bottomTower or part == ENUM.WallPart.upperTower then return false end
		if battle:getWallState(part) ~= nil and battle:getWallState(part) > 0 then return false end
	end
	return true
end

function Script:applicableGeneral(mechanics, problem)
	if self.hidden and self.hideNative then
		local battle = mechanics:getBattle()
		if battle:hasNativeStack(otherSide(mechanics:getCasterSide())) then
			problem:addStandard(mechanics, ENUM.SpellCastProblem.noAppropriateTarget)
			return false
		end
	end
	return true
end

local function noRoomToPlace(mechanics, problem)
	problem:addStandard(mechanics, ENUM.SpellCastProblem.noAppropriateTarget)
	return false
end

function Script:applicableTarget(mechanics, problem, target)
	if mechanics:isMassive() then return true end

	if #target == 0 then return noRoomToPlace(mechanics, problem) end

	local battle = mechanics:getBattle()
	local opts   = sideOptions(self, mechanics:getCasterSide())
	local shapes = shapesFor(opts)

	for _, dest in ipairs(target) do
		for _, shape in ipairs(shapes) do
			local hex = applyShape(dest.hex, shape)
			if not isHexAvailable(battle, hex, true) then
				return noRoomToPlace(mechanics, problem)
			end
		end
	end
	return true
end

function Script:transformTarget(mechanics, aimPoint, spellTarget)
	if mechanics:isMassive() then return {} end

	local opts   = sideOptions(self, mechanics:getCasterSide())
	local ranges = rangesFor(opts)

	local result = {}
	for _, dest in ipairs(spellTarget) do
		for _, range in ipairs(ranges) do
			local hex = applyShape(dest.hex, range)
			result[#result+1] = { hex = hex }
		end
	end
	return result
end

function Script:adjustAffectedHexes(mechanics, hexes, spellTarget)
	local transformed = self:transformTarget(mechanics, spellTarget, spellTarget)
	local opts   = sideOptions(self, mechanics:getCasterSide())
	local shapes = shapesFor(opts)
	for _, dest in ipairs(transformed) do
		for _, shape in ipairs(shapes) do
			local hex = applyShape(dest.hex, shape)
			if hex:isValid() then
				hexes:insert(hex)
			end
		end
	end
end

local function buildDescriptor(self, mechanics, side, hex, customSize)
	local spell = mechanics:getSpell()
	local opts  = sideOptions(self, side)
	return {
		pos              = hex,
		obstacleType     = ENUM.ObstacleType.spellCreated,
		spell            = spell,
		casterSpellPower = mechanics:getEffectPower(),
		spellLevel       = mechanics:getEffectLevel(),
		casterSide       = side,
		turnsRemaining   = self.turnsRemaining or -1,
		hidden           = self.hidden or false,
		passable         = self.passable or false,
		nativeVisible    = not (self.hideNative or false),
		trap             = self.trap or false,
		removeOnTrigger  = self.removeOnTrigger or false,
		trigger          = self.triggerAbility or "",
		appearSound      = opts.appearSound or "",
		appearAnimation  = opts.appearAnimation or "",
		animation        = opts.animation or "",
		customSize       = customSize,
	}
end

local function collectAvailable(battle, hexes)
	local out = {}
	for _, hex in ipairs(hexes) do
		if isHexAvailable(battle, hex, true) then
			out[#out+1] = hex
		end
	end
	return out
end

local function hexesFromArray(arr)
	local list = {}
	for i = 1, arr:size() do
		list[#list+1] = arr:at(i)
	end
	return list
end

function Script:apply(mechanics, server, target)
	local battle    = mechanics:getBattle()
	local side      = mechanics:getCasterSide()
	local opts      = sideOptions(self, side)
	local shapes    = shapesFor(opts)
	local patchCount = self.patchCount or 0

	local destinations = {}

	if patchCount > 0 then
		local candidates
		if mechanics:isMassive() then
			candidates = hexesFromArray(battle:getAllPossibleHexes())
		else
			candidates = {}
			for _, dest in ipairs(target) do
				candidates[#candidates+1] = dest.hex
			end
		end

		local available = collectAvailable(battle, candidates)

		for i = #available, 2, -1 do
			local j = server:rngInt(1, i)
			available[i], available[j] = available[j], available[i]
		end

		local toPlace = math.min(patchCount, #available)
		for i = 1, toPlace do
			destinations[#destinations+1] = available[i]
		end
	else
		for _, dest in ipairs(target) do
			destinations[#destinations+1] = dest.hex
		end
	end

	for _, hex in ipairs(destinations) do
		local customSize = {}
		for _, shape in ipairs(shapes) do
			local shaped = applyShape(hex, shape)
			customSize[#customSize+1] = shaped
		end
		local descriptor = buildDescriptor(self, mechanics, side, hex, customSize)
		server:addObstacle(battle, descriptor)
	end
end

return Script
